core/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -0,0 +1,177 @@
+    public static final String DEFAULT_FIELD = "_uid";
+
+    private String field = DEFAULT_FIELD;
+    private int id = -1;
+    private int max = -1;
+
+    public SliceBuilder() {
+    }
+
+    public SliceBuilder(StreamInput in) throws IOException {
+        this.field = in.readOptionalString();
+        this.id = in.readVInt();
+        this.max = in.readVInt();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalString(field);
+        out.writeVInt(id);
+        out.writeVInt(max);
+    }
+
+    public SliceBuilder setField(String field) {
+        if (field != null) {
+            this.field = field;
+        }
+        return this;
+    }
+
+    public SliceBuilder setId(int id) {
+        if (id < 0) {
+            throw new IllegalArgumentException("id must be greater than or equal to 0");
+        }
+        if (max != -1 && id >= max) {
+            throw new IllegalArgumentException("max must be greater than id");
+        }
+        this.id = id;
+        return this;
+    }
+
+    public SliceBuilder setMax(int max) {
+        if (max <= 0) {
+            throw new IllegalArgumentException("max must be greater than 0");
+        }
+        if (id != -1 && id >= max) {
+            throw new IllegalArgumentException("max must be greater than id");
+        }
+        this.max = max;
+        return this;
+    }
+
+    public SliceQuery toFilter(QueryShardContext context) {
+        final MappedFieldType type = context.fieldMapper(field);
+        if (type == null) {
+            throw new IllegalArgumentException("Field " + field + " not found");
+        }
+        if (UidFieldMapper.NAME.equals(field)) {
+            return new TermsSliceQuery(context.getIndexReader().getCoreCacheKey(), new Slice(field, id, max));
+        } else if (type.hasDocValues()) {
+            IndexFieldData<?> indexFieldData = context.getForField(type);
+            return new DocValuesSliceQuery(new Slice(field, id, max), indexFieldData);
+        }
+        throw new IllegalArgumentException("Can't slice on [" + field + "] because doc_values are not enabled");
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        innerToXContent(builder);
+        builder.endObject();
+        return builder;
+    }
+
+    void innerToXContent(XContentBuilder builder) throws IOException {
+        builder.field("field", field);
+        builder.field("id", id);
+        builder.field("max", max);
+    }
+
+    public static SliceBuilder fromXContent(QueryParseContext context) throws IOException {
+        XContentParser parser = context.parser();
+        XContentParser.Token token;
+        String currentFieldName = null;
+        String field = null;
+        int id = -1;
+        int max = -1;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
we should use `org.elasticsearch.common.xcontent.ObjectParser` for this instead of parsing all the stuff ourself